### PR TITLE
QE: Add smoke tests for Grafana in BV

### PR DESCRIPTION
**Pull request by vandabarata**
## What does this PR change?

Adds tests in the BV testsuite for the Grafana formula
- Use `monitoring_server` as monitoring server (for Prometheus _and_ Grafana)
- Access the grafana endpoint to make sure Grafana is working
- Open Grafana and confirm the 4 expected default dashboards are listed

## GUI diff
No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- Cucumber tests were added for BV teststuite

- [x] **DONE**

## Links
Closes https://github.com/SUSE/spacewalk/issues/17627
Ports:
- 4.2 https://github.com/SUSE/spacewalk/pull/19957
- 4.3 https://github.com/SUSE/spacewalk/pull/19891
- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests" (Test skipped, there are no changes to test)
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template [hunk 1]
```diff
@@ -137,8 +137,36 @@ Feature: Smoke tests for <client>
...
+    And I check "node" exporter
+    And I check "apache" exporter
+    And I check "postgres" exporter
+    And I click on "Save"
+    Then I should see a "Formula saved" text
+    # Apply highstate for Prometheus exporters
+    When I follow "States" in the content area
+    And I click on "Apply Highstate"
+    Then I should see a "Applying the highstate has been scheduled." text
+    And I wait until event "Apply highstate scheduled by admin" is completed
```
Comment (Bischoff) [inline testsuite/features/build_validation/smoke_tests/smoke_tests.template:159]: when

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template [hunk 2]
```diff
@@ -137,8 +137,36 @@ Feature: Smoke tests for <client>
...
+    # Visit monitoring endpoints on the minion
+    When I wait until "node" exporter service is active on "<client>"
+    And I visit "Prometheus node exporter" endpoint of this "<client>"
```
Comment (Bischoff) [inline testsuite/features/build_validation/smoke_tests/smoke_tests.template:162]: these are tests that it works.

`Then the "Prometheus node exporter" of this "<client>" should answer`

**Review by Bischoff — CHANGES_REQUESTED** [review]
some nitpickings

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template (hunk 2 above)
Comment (vandabarata) [reply testsuite/features/build_validation/smoke_tests/smoke_tests.template:162]: this wasn't added by me, it already existed, and all monitoring tests use this step :sweat_smile: should I rephrase this anyway?

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 3]
```diff
@@ -1109,3 +1110,8 @@
 Then(/^I should see the text "(.*?)" in the (Operating System|Architecture|Channel Label) field/) do |text, field|
   page.has_field?(text, with: field)
 end
+
+When(/^I visit the grafana dashboards of this "([^"]*)"$/) do |host|
```
[the remark this answers is not in the document]
Comment (vandabarata) [reply testsuite/features/step_definitions/navigation_steps.rb:1114]: sorry, I didn't get the suggestion here :grimacing:

[the remark this answers is not in the document]
Comment (Bischoff) [reply testsuite/features/step_definitions/navigation_steps.rb:1114]:
```
Then(/^the "Prometheus node exporter" of this ""([^"]*)" should answer$/) do |host|
```

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template (hunk 2 above)
Comment (Bischoff) [reply testsuite/features/build_validation/smoke_tests/smoke_tests.template:162]: ??? I can see it added in the steps file below

**Comment by srbarrios** [comment]
I think that's not the correct approach sorry.

I think we need to have as infrastructure:

- SUSE Manager Server
- SUSE Manager Proxy
- Monitoring Server (must have Prometheus and Grafana salt formulas)
- All our minions (must have the Prometheus Exporter salt formula)

We first prepare a SUSE minion that will act as our Monitoring server, applying the prometheus and grafana formulas.

Then we test every <client>, by applying the Prometheus exporter formula and checking that the entry of this minion is available in the dashboard from the monitoring server.

In addition, but a different scenario, we should also check in all our suse minions, if the Prometheus and Grafana salt formulas can work on them, applying it, and visiting the dashboards.

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template (hunk 2 above)
Comment (vandabarata) [reply testsuite/features/build_validation/smoke_tests/smoke_tests.template:162]: but it wasn't, I just moved it. I didn't change anything in the exporter formulas test, just moved it up so I'd have the monitoring server tests all together

**Comment by vandabarata** [comment]
Depends on https://github.com/uyuni-project/sumaform/issues/1207 to be done first

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template [hunk 4]
```diff
@@ -138,8 +138,36 @@ Feature: Smoke tests for <client>
...
+    And I wait until "apache" exporter service is active on "<client>"
+    And I visit "Prometheus apache exporter" endpoint of this "<client>"
+    And I wait until "postgres" exporter service is active on "<client>"
+    And I visit "Prometheus postgres exporter" endpoint of this "<client>"
+
+@monitoring_server
+  Scenario: Test <client> as Monitoring Server using Prometheus formula
```
Comment (srbarrios) [inline testsuite/features/build_validation/smoke_tests/smoke_tests.template:170]: We need to remove completely this scenarion as agreed this morning.

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template [hunk 5]
```diff
@@ -159,33 +187,35 @@ Feature: Smoke tests for <client>
     When I wait until "prometheus" service is active on "<client>"
     And I visit "Prometheus" endpoint of this "<client>"
 
-  Scenario: Test Prometheus exporter formula on the <client>
+@monitoring_server
+  Scenario: Test <client> as Monitoring Server using Grafana formula
```
Comment (srbarrios) [inline testsuite/features/build_validation/smoke_tests/smoke_tests.template:191]: We need to remove completely this scenario as agreed this morning

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 6]
```diff
@@ -1126,3 +1127,8 @@
     end
   end
 end
+
+When(/^I visit the grafana dashboards of this "([^"]*)"$/) do |host|
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb:1131]: There is no need ot this step.

What we need is a step visiting the grafana running in the monitoring server, not in the client.

And then, another step checking that the dashboard show a graph for a concrete client.

File: testsuite/features/build_validation/init_clients/monitoring_server.feature [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+@proxy
+  Scenario: Check registration on proxy of Monitoring Server
+    Given I am on the Systems overview page of this "proxy"
+    When I follow "Details" in the content area
+    And I follow "Proxy" in the content area
+    Then I should see "monitoring_server" hostname
+
+  Scenario: Check events history for failures on Monitoring Server
+    Given I am on the Systems overview page of this "monitoring_server"
+    Then I check for failed events on history event page
```
Comment (srbarrios) [inline testsuite/features/build_validation/init_clients/monitoring_server.feature:47]: Do you miss here the scenarios to apply the prometheus and grafana salt formulas to our monitoring server. And after that, check that prometheus service is up and running and same for grafana.

**Review by srbarrios — CHANGES_REQUESTED** [review]

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template (hunk 4 above)
Comment (vandabarata) [reply testsuite/features/build_validation/smoke_tests/smoke_tests.template:170]: this is just being applied on the monitoring server though - not on any other client, hence the tag

File: testsuite/features/build_validation/smoke_tests/smoke_tests.template (hunk 5 above)
Comment (vandabarata) [reply testsuite/features/build_validation/smoke_tests/smoke_tests.template:191]: same as above, only the monitoring server

File: testsuite/features/build_validation/init_clients/monitoring_server.feature (hunk 7 above)
Comment (vandabarata) [reply testsuite/features/build_validation/init_clients/monitoring_server.feature:47]: That's being done already, in the smoke tests - this is just to "start" the monitoring server minion and make sure it's working normally

File: testsuite/documentation/cucumber-steps.md [hunk 8]
```diff
@@ -30,17 +30,18 @@ The corresponding notion in Cucumber steps is "step host names".
...
+| Test host                       | Ruby target              | Bash environment variable                                | Step host name           | sumaform module            |
+|---------------------------------|--------------------------|----------------------------------------------------------|--------------------------|----------------------------|
+| Uyuni server                    | ```$server```            | ```$SERVER```                                            |                          | ```"suse_manager"```       |
+| Uyuni proxy                     | ```$proxy```             | ```$PROXY```                                             | ```"proxy"```            | ```"suse_manager_proxy"``` |
+| SLES Salt minion                | ```$minion```            | ```$MINION```                                            | ```"sle_minion"```       | ```"minion"```             |
+| SLES Docker and Kiwi build host | ```$build_host```        | ```$BUILD_HOST```                                        | ```"build_host"```       | ```"minion"```             |
```
Comment (Bischoff) [inline testsuite/documentation/cucumber-steps.md:38]: sumaform module is 'build_host", sorry i forgot to change it here

**Review by Bischoff — APPROVED** [review]
see requested nitpicks, otherwise lgtm

**Review by srbarrios — APPROVED** [review]

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 9]
```diff
@@ -995,15 +995,26 @@
 When(/^I visit "([^"]*)" endpoint of this "([^"]*)"$/) do |service, host|
   node = get_target(host)
   system_name = get_system_name(host)
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb:997]: Check if you still use this variable.

**Review by srbarrios — APPROVED** [review]

File: testsuite/features/step_definitions/navigation_steps.rb [hunk 10]
```diff
@@ -1014,7 +1014,8 @@
 end
 
 When(/^I enter the "([^"]*)" hostname as the Prometheus URL$/) do |host|
-  step %(I enter "http://#{host.full_hostname}" as "Prometheus URL")
+  node = get_target(host)
+  step %(I enter "http://#{node.full_hostname}" as "Prometheus URL")
 end
```
Comment (srbarrios) [inline testsuite/features/step_definitions/navigation_steps.rb:1019]: Maybe we can make this wrapper generic and can be useful in other scenarios.

```
When(/^I enter the "([^"]*)" hostname as "([^"]*)"$/) do |host, field|
  node = get_target(host)
  step %(I enter "http://#{node.full_hostname}" as "#{field}")
end
```

File: testsuite/features/support/twopence_init.rb [hunk 11]
```diff
@@ -72,6 +72,7 @@ def twopence_init(target)
   $opensuse154arm_ssh_minion = twopence_init("ssh:#{ENV['OPENSUSE154ARM_SSHMINION']}") if ENV['OPENSUSE154ARM_SSHMINION']
   $sle12sp5_buildhost = twopence_init("ssh:#{ENV['SLE12SP5_BUILDHOST']}") if ENV['SLE12SP5_BUILDHOST']
   $sle15sp4_buildhost = twopence_init("ssh:#{ENV['SLE15SP4_BUILDHOST']}") if ENV['SLE15SP4_BUILDHOST']
+  $monitoring_server = twopence_init("ssh:#{ENV['MONITORING_SERVER']}") if ENV['MONITORING_SERVER']
```
Comment (nodeg) [inline testsuite/features/support/twopence_init.rb:75]: Do you already adjusted the `.bashrc` on the controller in sumaform and added those environment variable?

**Review by nodeg — APPROVED** [review]

File: testsuite/features/support/twopence_init.rb (hunk 11 above)
Comment (vandabarata) [reply testsuite/features/support/twopence_init.rb:75]: @Bischoff did! https://github.com/uyuni-project/sumaform/pull/1223

Comment (nodeg) [reply testsuite/features/support/twopence_init.rb:75]: Alright. All good then :wink:

**Pull request merged**